paderbox/array/kernel.py

```diff
@@ -0,0 +1,212 @@
+    >>> _plot(a); _plot(np_kernel1d(a, 3, kernel=np.mean))
+    '    ████    '
+    '   ▃▅██▅▃   '
+
+    """
+    assert kernel_size % 2 == 1, (kernel_size, 'kernel size has to be odd.')
+    assert pad_position in ['pre', 'post', None], pad_position
+
+    if pad_position == 'pre':
+        shift = kernel_size // 2
+        padding = [(0, 0)] * (x.ndim - 1) + [(shift, shift)]
+        x = np.pad(x, padding, padding_mode)
+
+    y = kernel(pb.array.segment_axis(x, kernel_size, 1, end='pad'), axis=-1)
+
+    if pad_position == 'post':
+        shift = kernel_size // 2
+        padding = [(0, 0)] * (y.ndim - 1) + [(shift, shift)]
+        y = np.pad(y, padding, padding_mode)
+    return y
+
+
+def _ai_dilate_erode(ai, khalf):
+    """
+
+
+    # Test inverse mode an unknown shape
+    >>> a = pb.array.interval.ones()
+    >>> a[3:7] = False
+    >>> _plot(a[:10]); _ = [_plot(ai_dilate(a, k)[:10]) for k in [1, 3, 5, 11]]
+    '███    ███'
+    '███    ███'
+    '████  ████'
+    '██████████'
+    '██████████'
+
+    """
+    pairs = np.array(ai.normalized_intervals)
+    assert pairs.shape[-1] == 2, pairs.shape
+    if ai.inverse_mode:
+        khalf = -khalf
+    pairs[:, 0] -= khalf
+    pairs[:, 1] += khalf
+    pairs = np.maximum(pairs, 0)
+    if ai.shape is not None:
+        pairs = np.minimum(pairs, ai.shape[-1])
+    return pb.array.interval.core.ArrayInterval.from_pairs(
+        pairs.tolist(), shape=ai.shape, inverse_mode=ai.inverse_mode)
+
+
+def ai_dilate(ai, kernel_size):
+    """
+    >>> import paderbox as pb
+    >>> a = pb.array.interval.ArrayInterval(np.array([0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,], dtype=bool))
+    >>> a
+    ArrayInterval("4:8", shape=(12,))
+    >>> _plot(a); _ = [_plot(ai_dilate(a, k)) for k in [1, 3, 5, 11]]
+    '    ████    '
+    '    ████    '
+    '   ██████   '
+    '  ████████  '
+    '████████████'
+
+
+    """
+    assert kernel_size % 2 == 1, (kernel_size, 'kernel size has to be odd.')
+    return _ai_dilate_erode(ai, kernel_size//2)
+
+
+def ai_erode(ai, kernel_size):
+    """
+
+    >>> import paderbox as pb
+    >>> a = pb.array.interval.ArrayInterval(np.array([0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,], dtype=bool))
+    >>> a
+    ArrayInterval("4:8", shape=(12,))
+    >>> _plot(a); _ = [_plot(ai_erode(a, k)) for k in [1, 3, 11]]
+    '    ████    '
+    '    ████    '
+    '     ██     '
+    '            '
+    """
+    assert kernel_size % 2 == 1, (kernel_size, 'kernel size has to be odd.')
+    return _ai_dilate_erode(ai, -(kernel_size//2))
+
+
+def max_kernel1d(x, kernel_size):
+    """
+    Apply a max kernel to x.
+    In case of a boolean arrays, this operation is known as dilation.
+
+    >>> a = np.array([0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,])
+    >>> _plot(a); _plot(max_kernel1d(a, 3))
+    '    ████    '
+    '   ██████   '
+    """
+    if isinstance(x, np.ndarray):
+        return np_kernel1d(x, kernel_size, kernel=np.amax)
+    elif hasattr(x, 'normalized_intervals'):
```

Comment:
Why no `isinstance` check?

Reply:
I don't want to trigger the import. ArrayInterval is cython.